celements-model/src/main/java/com/celements/model/util/DefaultModelUtils.java

```diff
@@ -82,31 +93,55 @@ public EntityReference resolveRef(String name, EntityReference baseRef) {
       throw new IllegalArgumentException("name may not be empty");
     } else if (type == getRootEntityType()) {
       // resolver cannot handle root reference
-      resolvedRef = new WikiReference(name);
+      resolvedRef = RefBuilder.create().wiki(name).build(WikiReference.class);
     } else {
       baseRef = References.combineRef(baseRef, context.getWikiRef()).get();
       resolvedRef = resolver.resolve(name, type, baseRef);
     }
-    return cloneRef(resolvedRef, token); // effective immutability
+    return References.cloneRef(resolvedRef, token); // effective immutability
   }
 
   @Override
   public WikiReference getMainWikiRef() {
-    return new WikiReference("xwiki");
+    return RefBuilder.create().wiki("xwiki").build(WikiReference.class);
+  }
+
+  @Override
+  public Stream<WikiReference> getAllWikis() {
+    Stream<WikiReference> stream;
+    try {
+      stream = getXWiki().getVirtualWikisDatabaseNames(context.getXWikiContext()).stream()
+          .map(name -> RefBuilder.create().wiki(name).build(WikiReference.class));
+    } catch (XWikiException xwe) {
+      stream = Stream.of(context.getWikiRef());
+    }
+    return Stream.concat(Stream.of(getMainWikiRef()), stream).distinct();
+  }
+
+  @Override
+  public Stream<SpaceReference> getAllSpaces(WikiReference wikiRef) {
+    RefBuilder builder = RefBuilder.from(wikiRef);
+    try {
+      return ContextExecutor.executeInWikiThrows(wikiRef,
+          () -> getXWiki().getSpaces(context.getXWikiContext()).stream())
```

Comment:
please inline "getSpaces" to eliminate the sensless XWikiException wrapper and mark the "getSpaces" method in XWiki.java as deprecated.